iterative/utils/logger.go

```diff
@@ -0,0 +1,104 @@
+package utils
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/sirupsen/logrus"
+	"strings"
+	"time"
+)
+
+var baseTimestamp = time.Now()
+var colors = make(map[string]int)
+
+type basicFormatter struct{}
+
+func (f *basicFormatter) Format(entry *logrus.Entry) ([]byte, error) {
+	levelText := strings.ToUpper(entry.Level.String())
+	levelColor := colors[levelText]
+	tpl := "[%s] 🚀\x1b[%dmTPI\x1b[0m %s\n"
+	return []byte(fmt.Sprintf(tpl, levelText, levelColor, entry.Message)), nil
+}
+
+func init() {
+	colors["DEBUG"] = 36
+	colors["INFO"] = 36
+	colors["WARN"] = 33
+	colors["ERROR"] = 31
+	colors["FATAL"] = 31
+	colors["purple"] = 35
+
+	logrus.SetLevel(logrus.DebugLevel)
+	logrus.SetFormatter(&basicFormatter{})
+}
+
+type tpiFormatter struct{}
+
+func (f *tpiFormatter) Format(entry *logrus.Entry) ([]byte, error) {
+	data := make(logrus.Fields)
+	for k, v := range entry.Data {
+		data[k] = v
+	}
+
+	d := data["d"].(*schema.ResourceData)
+	message := entry.Message
+	levelText := strings.ToUpper(entry.Level.String())
+	levelColor := colors[levelText]
+
+	if message == "instance" {
+		cloud := d.Get("cloud").(string)
+		machine := d.Get("machine").(string)
+		region := d.Get("region").(string)
+		spot := d.Get("spot").(float64)
+
+		spottext := ""
+		if spot > 0 {
+			spottext = fmt.Sprintf("(Spot %f/h)", spot)
+		}
+		message = fmt.Sprintf("🚀 %s %s%s at %s", cloud, machine, spottext, region)
+	}
+
+	if message == "status" {
+		status := d.Get("status").(map[string]interface{})
+
+		running := "not yet started"
+		if status["running"] != nil {
+			running = "is terminated"
+			if status["running"].(int) == 1 {
+				running = "is running 🟡"
+			}
+		}
+
+		success := ""
+		if running == "is terminated" {
+			success = "without any output"
+			if status["succeeded"] != nil && status["succeeded"].(int) == 1 {
+				success = "succesfully 🟢"
+			}
+			if status["failed"] != nil && status["failed"].(int) == 1 {
+				success = "with errors 🔴"
+			}
+		}
+
+		message = fmt.Sprintf("Task %s %s", running, success)
+	}
+
+	if message == "logs" {
+		logs := d.Get("logs").([]interface{})
+		taskLogs := "No logs"
+		if len(logs) > 0 {
+			taskLogs = strings.Replace(logs[0].(string), "\n", fmt.Sprintf("\n[%s] ", levelText), -1)
```

Comment:
```suggestion
			taskLogs = strings.Replace(logs[0].(string), "\n", fmt.Sprintf("\n%s:", levelText), -1)
```

Reply:
This is needing to align internal TF logs with the logrus level

Reply:
Sounds [chaotic neutral versus lawful evil](https://xkcd.com/2408/). 🤔 Applying this suggestion would make some `logrus` log lines look different to the standard library `log` ones.

```diff
- 2022-03-03T17:22:29.743Z [INFO]  provider.terraform-provider-iterative: 2022/03/03 17:22:29 [INFO] Reading ...
+ 2022-03-03T17:22:29.743Z [INFO]  provider.terraform-provider-iterative: 2022/03/03 17:22:29 INFO Reading ...
```

Still, this drawback doesn't seem to be relevant anymore after the changes introduced with cb5d9ec35be30e4d19f38fa23ee72b4f2af3db7b. Now all our user–facing logs use `logrus` and we can choose what format to apply.

Reply:
I'd consider removing the brackets and sparing another two bytes from the required terminal width. 🤷🏼‍♂️ Still, it's probably bikeshedding at its finest.

Reply:
The main thing is we shouldn't have datetime & level appear twice on a line, and should replace `provider.terraform-provider-iterative` with `TPI`.

Reply:
I would also like to have a grand piano under my bed. https://github.com/iterative/terraform-provider-iterative/pull/423#issuecomment-1060592746

Reply:
We can definitely drop the second datetime, right?

Reply:
Yes, @DavidGOrtega already did that. See https://github.com/iterative/terraform-provider-iterative/pull/423#issuecomment-1058573507. Still, machine logs will necessarily need to have a different datetime, because what matters it the time when the log was collected, and not the time when it was displayed to the user.